src/core/bpf-lsm.c

```diff
@@ -0,0 +1,260 @@
+                return supported = 0;
+        }
+
+        r = bpf_object_load(obj);
+        if (r < 0) {
+                log_debug_errno(errno, "Failed to load LSM BPF program: %m");
+                return supported = 0;
+        }
+
+        return supported = 1;
+}
+
+int lsm_bpf_setup(void) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        struct bpf_program *prog = NULL;
+        struct bpf_link *link = NULL;
+        int outer_map_fd, inner_map_fd;
+        int r;
+
+        r = bpf_object_new(restrict_fs_hexdump_buffer, sizeof(restrict_fs_hexdump_buffer), &obj);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create BPF object from hexdump buffer: %m");
+
+        /* Dummy map to satisfy the verifier */
+        inner_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(uint32_t), sizeof(uint32_t), 128, 0);
+        if (inner_map_fd < 0)
+                return log_debug_errno(errno, "Failed to create BPF map");
+
+        r = bpf_object_set_inner_map_fd(obj, "cgroup_hash", inner_map_fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set inner map fd: %m");
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to load LSM BPF program: %m");
+
+        /* Close dummy map */
+        close(inner_map_fd);
+
+        prog = bpf_object__find_program_by_title(obj, LSM_PROG_SECTION_NAME);
+        if (!prog) {
+                log_error("Failed to find LSM BPF program '%s'", LSM_PROG_SECTION_NAME);
+                return -EINVAL;
+        }
+
+        link = bpf_program__attach_lsm(prog);
+        if (!link)
+                return log_error_errno(errno, "Failed to attach BPF LSM program: %m");
+
+        log_info("LSM BPF program attached");
+
+        outer_map_fd = bpf_object_get_map_fd(obj, "cgroup_hash");
+        if (outer_map_fd < 0)
+                return log_error_errno(outer_map_fd, "Failed to get map fd: %m");
+
+        r = mkdir_parents(LSM_MAP_BPFFS_PATH, 0700);
+        if (r < 0)
+                return log_error_errno(r, "Failed create systemd bpffs directory: %m");
+
+        r = bpf_obj_pin(outer_map_fd, LSM_MAP_BPFFS_PATH);
+        if (r < 0)
+                return log_error_errno(r, "Failed to pin LSM BPF map in the bpffs: %m");
+
+        return 0;
+}
+
+int bpf_restrict_filesystems(char **filesystems, char *cgroup_path) {
+        _cleanup_close_ int inner_map_fd = -1, outer_map_fd = -1;
+        _cleanup_free_ char *path = NULL;
+        uint64_t cgroup_id;
+        uint32_t magic, dummy_value = 1;
+        char **f;
+        int r;
+
+        assert(filesystems);
+        assert(cgroup_path);
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, cgroup_path, NULL, &path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get systemd cgroup path");
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
```

Comment:
Before calling this function in execute.c, `lsm_bpf_supported()` is called which checks cgroup2 is supported so this will never be called in legacy mode. Do you think I should add a check in this function too?

I'll update the PR description and perhaps add this to the documentation :+1: